### PR TITLE
refactor(memory): extract shared batch embedding lifecycle

### LINKED ISSUE #31155: [Refactor]: Extract shared batch embedding lifecycle from OpenAI/Voyage/Gemini providers
### Summary

Extract the duplicated poll/wait lifecycle from the three batch embedding providers into a single shared `waitForBatch()` function.

### Problem to solve

The three batch embedding provider files (`batch-openai.ts`, `batch-voyage.ts`, `batch-gemini.ts`) contain ~903 LOC total with 5 structurally identical functions copy-pasted across each provider. The `waitFor*Batch` poll loops are nearly character-for-character identical — differing only in provider name strings, terminal state constants, and output file ID extraction.

This duplication has already caused a real bug: issue #15738 (Gemini Batch Embedding Infinite Polling Loop) was caused by the Gemini wait loop using different terminal-state names (`SUCCEEDED`/`COMPLETED`/`DONE` vs `completed`) than OpenAI/Voyage, with no shared implementation to ensure consistency. Making the batch timeout configurable (#25157) would require changes in all 3 files instead of 1. Error handling for silent 404 failures (#29568) faces the same per-provider maintenance burden.

Multiple prior refactoring attempts ("consolidate embeddings and batch helpers", "dedupe batch helpers", "share batch create retry", "share batch output parsing") extracted the outer runner/grouping logic (`batch-runner.ts`, `batch-utils.ts`, `batch-output.ts`) but left each provider to independently re-implement the core submit→poll→wait→timeout lifecycle.

### Proposed solution

Extract a generic `waitForBatch<TStatus>()` function into `src/memory/batch-lifecycle.ts` parameterized by a `BatchLifecycleAdapter<TStatus>` record. Each provider supplies a small adapter with: `fetchStatus`, `resolveState`, `isCompleted`, `isFailed`, `resolveOutputFileId`, and optional `resolveErrorDetail`. The shared function handles the poll/wait/timeout loop. Each provider file shrinks by ~70-80 LOC, and the total net reduction is ~210 LOC replaced by ~95 LOC of shared code.

### Alternatives considered

A class hierarchy was considered but rejected — the adapter record pattern is lighter, requires no inheritance, and matches the existing `createRemoteEmbeddingProvider` pattern already proven in the codebase.

### Impact

- **Affected**: `src/memory/batch-openai.ts`, `src/memory/batch-voyage.ts`, `src/memory/batch-gemini.ts` — every batch embedding operation
- **Severity**: Active bugs from divergent implementations (#15738)
- **Frequency**: Every batch embedding operation uses this code path
- **Consequence**: Bug fixes must be applied 3x, terminal state drift causes infinite loops

### Evidence/examples

The `waitForOpenAiBatch` (lines 134-179) and `waitForVoyageBatch` (lines 128-173) were character-for-character identical except for provider name strings. `waitForGeminiBatch` (lines 223-266) had the same structure but with divergent terminal state constants (`SUCCEEDED`/`COMPLETED`/`DONE` vs `completed`).

### Additional information

**New canonical abstractions:**
- `waitForBatch<TStatus>()` in `src/memory/batch-lifecycle.ts` — replaces `waitForOpenAiBatch`, `waitForVoyageBatch`, `waitForGeminiBatch`. Use for any batch embedding provider's poll/wait lifecycle. Documented in AGENTS.md under "Coding Style & Naming Conventions".

### PR DESCRIPTION
**Pull request by mcaxtr**
## Summary

- Problem: The three batch embedding providers (OpenAI, Voyage, Gemini) each independently implemented the same 5-step poll/wait lifecycle (~45 LOC each), with terminal state constants that had already diverged — directly causing #15738 (Gemini infinite polling loop).
- Why it matters: Every bug fix or improvement must be applied 3x. Adding configurable timeout (#25157) or fixing silent 404s (#29568) requires editing all 3 files. A new provider requires ~250 LOC of boilerplate.
- What changed: Extracted `waitForBatch<TStatus>()` into `src/memory/batch-lifecycle.ts` with a `BatchLifecycleAdapter` interface. Each provider now supplies a small adapter record (~15 LOC) instead of reimplementing the lifecycle. 10 new tests for the shared function. Net reduction: ~210 LOC removed, ~95 LOC added.
- What did NOT change (scope boundary): Provider-specific submit, fetch-status, fetch-output, and parse-output functions remain in their respective files. `batch-runner.ts` grouping/concurrency infrastructure unchanged. Config schema unchanged. No external API changes.

## Change Type (select all)

- [ ] Bug fix
- [ ] Feature
- [x] Refactor
- [x] Docs
- [ ] Security hardening
- [ ] Chore/infra

## Scope (select all touched areas)

- [ ] Gateway / orchestration
- [ ] Skills / tool execution
- [ ] Auth / tokens
- [x] Memory / storage
- [ ] Integrations
- [ ] API / contracts
- [ ] UI / DX
- [ ] CI/CD / infra

## Linked Issue/PR

- Closes #31155
- Related #15738, #25157, #29568

## User-visible / Behavior Changes

None. The batch embedding lifecycle behaves identically — same terminal states, same timeout, same error messages. The only observable difference: Gemini failure errors now consistently include `": unknown error"` suffix when `error.message` is absent (matching the original behavior that was verified during review).

## Security Impact (required)

- New permissions/capabilities? No
- Secrets/tokens handling changed? No
- New/changed network calls? No
- Command/tool execution surface changed? No
- Data access scope changed? No

### Environment

- OS: Linux
- Runtime/container: Node 22+ / Bun
- Model/provider: OpenAI, Voyage, Gemini (batch embedding providers)
- Relevant config: `remote.batch.*` settings

### Steps

1. Run `pnpm vitest run src/memory/batch-lifecycle` — 10 new tests for the shared lifecycle
2. Run `pnpm vitest run src/memory/batch-voyage` — 2 existing tests still pass
3. Run `pnpm build` — clean typecheck
4. Run `pnpm check` — clean lint/format

### Expected

- All tests pass, no type errors, no lint warnings

### Actual

- Confirmed: 12/12 tests pass, build clean, check clean

## Evidence

- [x] Failing test/log before + passing after
- [ ] Trace/log snippets
- [ ] Screenshot/recording
- [ ] Perf numbers (if relevant)

10 new tests in `batch-lifecycle.test.ts` covering: completed-on-first-poll, completed-after-polls, completed-without-output-file, failed-state, failed-with-error-detail, wait-disabled, timeout, debug-callback, initial-then-poll, label-in-error-messages. Plus 2 existing Voyage batch tests pass unchanged.

## Human Verification (required)

- Verified scenarios: Build passes, lint passes, all 12 tests pass (10 new lifecycle + 2 existing Voyage), codex review clean after fix round
- Edge cases checked: Initial status already completed (skip poll), timeout at 0ms, failed with and without error detail, wait disabled, unknown terminal states
- What you did **not** verify: Runtime batch embedding with real provider APIs (requires live keys and batch credits)

## Compatibility / Migration

- Backward compatible? Yes
- Config/env changes? No
- Migration needed? No

## Failure Recovery (if this breaks)

- How to disable/revert this change quickly: Revert the PR commit
- Files/config to restore: `src/memory/batch-openai.ts`, `src/memory/batch-voyage.ts`, `src/memory/batch-gemini.ts`
- Known bad symptoms reviewers should watch for: Batch embedding operations timing out unexpectedly, infinite polling loops, missing embeddings after batch completion

## Risks and Mitigations

- Risk: Provider API returns a new terminal state not in the adapter's `isCompleted`/`isFailed` set → infinite polling.
  - Mitigation: Same risk existed before (each provider had its own hardcoded sets). The shared lifecycle makes it easier to add new states in one place. The timeout mechanism is the safety net.

## New Canonical Abstractions (for future contributors)

- `waitForBatch<TStatus>()` in `src/memory/batch-lifecycle.ts` — single source of truth for batch embedding poll/wait lifecycle. Handles state detection, timeout, error extraction, and configurable polling via `BatchLifecycleAdapter`. Added to AGENTS.md under "Coding Style & Naming Conventions". Do not create per-provider wait loops.

### CODE REVIEW

File: src/memory/batch-lifecycle.ts [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+export async function waitForBatch<TStatus>(
+  params: WaitForBatchParams<TStatus>,
+): Promise<WaitForBatchResult> {
+  const { adapter, batchId, wait, debug } = params;
+  // Clamp poll/timeout to sane ranges to prevent API hammering or infinite waits.
+  const pollIntervalMs = Number.isFinite(params.pollIntervalMs)
+    ? Math.max(MIN_POLL_MS, params.pollIntervalMs)
+    : DEFAULT_POLL_MS;
+  const timeoutMs = Number.isFinite(params.timeoutMs)
+    ? Math.min(MAX_TIMEOUT_MS, Math.max(1, params.timeoutMs))
```
[the remark this answers is not in the document]
Comment (mcaxtr) [reply src/memory/batch-lifecycle.ts]: Good catch — dropped the 24h ceiling in f9a641a40. `timeoutMs` is now passed through with only a `Math.max(1, ...)` floor (matching the pre-refactor behavior where all three providers honored the caller value as-is). The `MIN_POLL_MS` floor on poll interval is retained to prevent API hammering.
